storm-core/src/jvm/org/apache/storm/daemon/supervisor/SyncProcessEvent.java

```diff
@@ -0,0 +1,672 @@
+
+        Boolean runWorkerAsUser = Utils.getBoolean(conf.get(Config.SUPERVISOR_RUN_WORKER_AS_USER), false);
+        String stormHome = ConfigUtils.concatIfNotNull(System.getProperty("storm.home"));
+        String stormOptions = ConfigUtils.concatIfNotNull(System.getProperty("storm.options"));
+        String stormConfFile = ConfigUtils.concatIfNotNull(System.getProperty("storm.conf.file"));
+        String workerTmpDir = ConfigUtils.workerTmpRoot(conf, workerId);
+
+        String stormLogDir = ConfigUtils.getLogDir();
+        String stormLogConfDir = (String) (conf.get(Config.STORM_LOG4J2_CONF_DIR));
+
+        String stormLog4j2ConfDir;
+        if (StringUtils.isNotBlank(stormLogConfDir)) {
+            if (Utils.isAbsolutePath(stormLogConfDir)) {
+                stormLog4j2ConfDir = stormLogConfDir;
+            } else {
+                stormLog4j2ConfDir = stormHome + Utils.FILE_PATH_SEPARATOR + stormLogConfDir;
+            }
+        } else {
+            stormLog4j2ConfDir = stormHome + Utils.FILE_PATH_SEPARATOR + "log4j2";
+        }
+
+        String stormRoot = ConfigUtils.supervisorStormDistRoot(conf, stormId);
+
+        String jlp = jlp(stormRoot, conf);
+
+        String stormJar = ConfigUtils.supervisorStormJarPath(stormRoot);
+
+        Map stormConf = ConfigUtils.readSupervisorStormConf(conf, stormId);
+
+        String workerClassPath = getWorkerClassPath(stormJar, stormConf);
+
+        Object topGcOptsObject = stormConf.get(Config.TOPOLOGY_WORKER_GC_CHILDOPTS);
+        List<String> topGcOpts = new ArrayList<>();
+        if (topGcOptsObject instanceof String) {
+            topGcOpts.add((String) topGcOptsObject);
+        } else if (topGcOptsObject instanceof List) {
+            topGcOpts.addAll((List<String>) topGcOptsObject);
+        }
+
+        int memOnheap = 0;
+        if (resources.get_mem_on_heap() > 0) {
+            memOnheap = (int) Math.ceil(resources.get_mem_on_heap());
+        } else {
+            //set the default heap memory size for supervisor-test
+            memOnheap = Utils.getInt(stormConf.get(Config.WORKER_HEAP_MEMORY_MB), 768);
+        }
+
+        int memoffheap = (int) Math.ceil(resources.get_mem_off_heap());
+
+        int cpu = (int) Math.ceil(resources.get_cpu());
+
+        List<String> gcOpts = null;
+
+        if (topGcOpts != null) {
+            gcOpts = substituteChildopts(topGcOpts, workerId, stormId, port, memOnheap);
+        } else {
+            gcOpts = substituteChildopts(conf.get(Config.WORKER_GC_CHILDOPTS), workerId, stormId, port, memOnheap);
+        }
+
+        Object topoWorkerLogwriterObject = stormConf.get(Config.TOPOLOGY_WORKER_LOGWRITER_CHILDOPTS);
+        List<String> topoWorkerLogwriterChildopts = new ArrayList<>();
+        if (topoWorkerLogwriterObject instanceof String) {
+            topoWorkerLogwriterChildopts.add((String) topoWorkerLogwriterObject);
+        } else if (topoWorkerLogwriterObject instanceof List) {
```

Comment:
checking instanceof List is not sufficient here as any collection is acceptable. You can do `instanceof Collection` and cast it to `Collection<String>` instead. Same can be done in other places as well.